src/core/eval.ml

```diff
@@ -192,9 +192,11 @@ and tree_walk : dtree -> ctxt -> stack -> (term * stack) option =
     let open Tree_types in
     match tree with
     | Fail                                                -> None
-    | Leaf(env_builder, act)                              ->
-        (* Allocate an environment for the action. *)
-        let env = Array.make (Bindlib.mbinder_arity act) TE_None in
+    | Leaf(env_builder, act, xvars)                       ->
+        (* Allocate an environment where to place terms coming from the
+           pattern variables for the action. *)
+        (* We have [length env_builder = mbinder_arity act - xvars] *)
+        let env = Array.make (Bindlib.mbinder_arity act - xvars) TE_None in
```

Comment:
Why not taking env of length mbinder_arity and filling the last xvars_nb variables by fresh metavariables (if xvars_nb <> 0 only for efficiency)? Then, msubst will not fail.